File: 0543/543.go
func diameterOfBinaryTree(root *TreeNode) int {
	if root == nil {
		return 0
	}
	l := diameterOfBinaryTree(root.Left)
	r := diameterOfBinaryTree(root.Right)
	if l > r {
		return 1 + l
	}
	return 1 + r
}

File: 0543/543.java

class Solution {
    int max = 0;
    public int diameterOfBinaryTree(TreeNode root) { depth(root); return max; }
    int depth(TreeNode node) {
        if (node == null) return 0;
        int l = depth(node.left), r = depth(node.right);
        max = Math.max(max, l + r);
        return 1 + Math.max(l, r);
    }
}
